pkg/api/openshiftclusterdocument.go

```diff
@@ -50,3 +52,15 @@ type OpenShiftClusterDocument struct {
 func (c *OpenShiftClusterDocument) String() string {
 	return encodeJSON(c)
 }
+
+func (doc *OpenShiftClusterDocument) ImageRegistryStorageAccountName() (string, error) {
```

Comment:
Do we actually need this function?  Can we just replace it with `doc.OpenShiftCluster.Properties.ImageRegistryStorageAccountName`? I don't see any issue with dereferencing the `doc` type or the `doc.OpenShiftCluster` type as they are already populated through cosmos.

Reply:
Honestly, I did it accessing the nested structs at the beginning but I decided to create this method because:

1- At that point, I felt I was violating the Law of Demeter -> from the perspective of the client talking to OpenShiftClusterDocument, I know it can provide me "something" called "image registry account name" but as I am interacting with that object, I don't want to know how to obtain it (information encapsulation).  What happens when you access the field using all the nested calls doc.OpenShiftCluster.Properties.ImageRegistryStorageAccountName you expose the details.

2- In order to access ImageRegistryStorageAccountName, we first need to ensure doc.OpenShiftCluster is not nil, and the benefit of having this function is that you just need to do the check inside the function. If you don't do it inside, you will need to do it N times where N is the number of calls to ImageRegistryStorageAccountName(). In that case, is just 1 but I don't see any drawback of having the function.

I know this type of nested access is common in Go, but for me, point 2 is important here to preserve encapsulation.

Have I convinced you?
If yes, I will directly resolve the conversation.
If not, I will apply your suggestion and resolve the conversation.

Thanks! 🙂

Reply:
I also think it is unnecessary

> I am interacting with that object, I don't want to know how to obtain it (information encapsulation). What happens when you access the field using all the nested calls doc.OpenShiftCluster.Properties.ImageRegistryStorageAccountName you expose the details.

`doc.OpenShiftCluster.Properties.ImageRegistryStorageAccountName` vs `doc.ImageRegistryStorageAccountName()` - doesn't make much difference.

Also following this logic - we should have a getter for every field we have on the `OpenShiftClusterDocument` level. Doesn't sound reasonable.

> In order to access ImageRegistryStorageAccountName, we first need to ensure doc.OpenShiftCluster is not nil

Also I don't think `doc.OpenShiftCluster` will ever be `nil`. I don't know why it is a pointer, but we set it straight away [on cluster creation](https://github.com/Azure/ARO-RP/blob/98fc429ecefc51d669cc72dc7cd3963231c18f1c/pkg/frontend/openshiftcluster_putorpatch.go#L68-L86) and never set to `nil`, I believe.

>  If you don't do it inside, you will need to do it N times where N is the number of calls to ImageRegistryStorageAccountName(). In that case, is just 1 but I don't see any drawback of having the function.

You still need to do `if err != nil` every time you call `ImageRegistryStorageAccountName`.